modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1562,6 +1562,19 @@ void GDScriptAnalyzer::resolve_assignable(GDScriptParser::AssignableNode *p_assi
 				parser->push_warning(p_assignable->initializer, GDScriptWarning::NARROWING_CONVERSION);
 #endif
 			}
+
+			if (p_assignable->initializer->is_constant && type_from_metatype(initializer_type).builtin_type != type_from_metatype(specified_type).builtin_type) {
+				Variant converted;
+				const Variant *value = &p_assignable->initializer->reduced_value;
+				Callable::CallError ce;
+				Variant::construct(specified_type.builtin_type, converted, &value, 1, ce);
+				if (ce.error == Callable::CallError::CALL_OK) {
+					p_assignable->initializer->reduced_value = converted;
+				} else {
+					// Normally, this should not happen.
+					ERR_PRINT(vformat(R"(Cannot convert "%s" initializer value of type %s to the specified type %s.)", p_assignable->identifier->name, initializer_type.to_string(), specified_type.to_string()));
+				}
+			}
```

Comment:
So now the conversion happens for every assignable (variables, constants, parameters) if their initializer is a constant expression. I think it is good. It means that there is no need to set `use_conversion_assign` in such cases because types will always match.

To get that benefit this `if` should merge with previous `if/else` thing, becoming its first `if`. Warning about `NARROWING_CONVERSION` should be decoupled from `if/else` construction and moved above it with condition including hard type check becoming - `specified_type.builtin_type == Variant::INT && initializer_type.is_hard_type() && initializer_type.builtin_type == Variant::FLOAT`.

Reply:
I did it a little differently: I moved the type incompatibility check above. I think this option is better. I also added the missing `Variant::can_convert_strict`.

But at the same time, I broke `const`-`preload`, since `is_type_compatible` for some reason does not work correctly with this (before that, the condition had `is_variable &&` part that hid this error).